libsolidity/codegen/ContractCompiler.cpp

```diff
@@ -522,91 +522,99 @@ bool ContractCompiler::visit(InlineAssembly const& _inlineAssembly)
+							errinfo_comment("Stack too deep, try removing local variables.")
+						);
+					for (unsigned i = 0; i < variable->type()->sizeOnStack(); ++i)
+						_assembly.append(dupInstruction(stackDiff));
 				}
-				else if (auto variable = dynamic_cast<VariableDeclaration const*>(decl))
+				else
 				{
-					solAssert(!variable->isConstant(), "");
-					if (m_context.isLocalVariable(variable))
+					solAssert(m_context.isStateVariable(variable), "Invalid variable type.");
+					auto const& location = m_context.storageLocationOfVariable(*variable);
+					if (!variable->type()->isValueType())
 					{
-						int stackDiff = _assembly.deposit() - m_context.baseStackOffsetOfVariable(*variable);
-						if (stackDiff < 1 || stackDiff > 16)
-							BOOST_THROW_EXCEPTION(
-								CompilerError() <<
-								errinfo_sourceLocation(_inlineAssembly.location()) <<
-								errinfo_comment("Stack too deep, try removing local variables.")
-							);
-						for (unsigned i = 0; i < variable->type()->sizeOnStack(); ++i)
-							_assembly.append(dupInstruction(stackDiff));
+						solAssert(location.second == 0, "Intra-slot offest assumed to be zero.");
+						_assembly.append(location.first);
 					}
 					else
 					{
-						solAssert(m_context.isStateVariable(variable), "Invalid variable type.");
-						auto const& location = m_context.storageLocationOfVariable(*variable);
-						if (!variable->type()->isValueType())
-						{
-							solAssert(location.second == 0, "Intra-slot offest assumed to be zero.");
-							_assembly.append(location.first);
-						}
-						else
-						{
-							_assembly.append(location.first);
-							_assembly.append(u256(location.second));
-						}
+						_assembly.append(location.first);
+						_assembly.append(u256(location.second));
 					}
 				}
-				else if (auto contract = dynamic_cast<ContractDefinition const*>(decl))
-				{
-					solAssert(contract->isLibrary(), "");
-					_assembly.appendLibraryAddress(contract->fullyQualifiedName());
-				}
-				else
-					solAssert(false, "Invalid declaration type.");
-			} else {
-				// lvalue context
-				auto variable = dynamic_cast<VariableDeclaration const*>(decl);
-				solAssert(
-					!!variable && m_context.isLocalVariable(variable),
-					"Can only assign to stack variables in inline assembly."
+			}
+			else if (auto contract = dynamic_cast<ContractDefinition const*>(decl))
+			{
+				solAssert(contract->isLibrary(), "");
+				_assembly.appendLibraryAddress(contract->fullyQualifiedName());
+			}
+			else
+				solAssert(false, "Invalid declaration type.");
+		} else {
+			// lvalue context
+			auto variable = dynamic_cast<VariableDeclaration const*>(decl);
+			solAssert(
+				!!variable && m_context.isLocalVariable(variable),
+				"Can only assign to stack variables in inline assembly."
+			);
+			unsigned size = variable->type()->sizeOnStack();
+			int stackDiff = _assembly.deposit() - m_context.baseStackOffsetOfVariable(*variable) - size;
```

Comment:
I see, `stackDiff` means "stack height difference between the new `v3` and the original `v3`" in the retrieve case.  `stackDiff` means "stack height difference between the stacktop and the topmost removed element" in the store case.